File: web/lib/track-edit-assets.ts
import crypto from "node:crypto";
import path from "node:path";

export const TRACK_EDIT_ASSET_ROOT_ENV = "TRACK_EDIT_ASSET_ROOT";

export function getTrackEditAssetRoot() {
  const configuredRoot = process.env[TRACK_EDIT_ASSET_ROOT_ENV]?.trim();
  if (configuredRoot) {
    return path.resolve(configuredRoot);
  }

  return path.resolve(process.cwd(), "storage", "track-edit-assets");
}

export function buildTrackCoverAssetKey(trackId: string, extension: string, basename = "cover") {
  const normalizedExtension = extension.startsWith(".") ? extension.toLowerCase() : `.${extension.toLowerCase()}`;
  const safeTrackId = trackId.replace(/[^a-zA-Z0-9_-]/g, "_");
  return path.posix.join(safeTrackId, `${basename}${normalizedExtension}`);
}

export function resolveTrackEditAssetPath(assetPathOrKey: string) {
  if (path.isAbsolute(assetPathOrKey)) {
    return path.normalize(assetPathOrKey);
  }

  return path.join(getTrackEditAssetRoot(), assetPathOrKey);
}

export function sha256Hex(input: Uint8Array) {
  return crypto.createHash("sha256").update(input).digest("hex");
}

File: worker/track_edit_assets.py
import os
from pathlib import Path


TRACK_EDIT_ASSET_ROOT_ENV = "TRACK_EDIT_ASSET_ROOT"


def get_track_edit_asset_root() -> Path:
    configured_root = os.environ.get(TRACK_EDIT_ASSET_ROOT_ENV, "").strip()
    if configured_root:
        return Path(configured_root).expanduser().resolve()

    workspace_web_root = Path("/workspace/web")
    if workspace_web_root.exists():
        return (workspace_web_root / "storage" / "track-edit-assets").resolve()

    repo_web_root = Path(__file__).resolve().parents[1] / "web"
    return (repo_web_root / "storage" / "track-edit-assets").resolve()


def build_track_cover_asset_key(track_id: str, extension: str, basename: str = "cover") -> str:
    normalized_extension = extension.lower() if extension.startswith(".") else f".{extension.lower()}"
    safe_track_id = "".join(char if char.isalnum() or char in {"_", "-"} else "_" for char in track_id)
    return f"{safe_track_id}/{basename}{normalized_extension}"


def resolve_track_edit_asset_path(asset_path_or_key: str | None) -> Path | None:
    if not asset_path_or_key:
        return None

    path = Path(asset_path_or_key)
    if path.is_absolute():
        return path

    return (get_track_edit_asset_root() / asset_path_or_key).resolve()
